File: questions/02_Add_Two_Numbers/code.py
class Solution(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def addTwoNumbers(self, l1, l2):
        dummy_head = ListNode(0)
        current = dummy_head
        carry = 0
    
        # Traverse both lists until both are None
        while l1 or l2 or carry:
            # Get the current values, if the node is None, use 0
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
        
            # Compute the sum and carry
            total = carry + x + y
            carry = total // 10
            current.next = ListNode(total % 10)
        
            # Move the current pointer forward
            current = current.next
        
            # Move to the next nodes in the lists if possible
            if l1: l1 = l1.next
            if l2: l2 = l2.next
    
        # Check if there's a leftover carry
        if carry > 0:
            current.next = ListNode(carry)
    
        return dummy_head.next

File: questions/25_reverse_nodes_in_k_group/code.js
var reverseKGroup = function(head, k) {
    const dummy = new ListNode();
    dummy.next = head;
    let prev = dummy
    
    //check if there are enough nodes
    while(true) {
        let node = prev;
        for (let i = 0; i < k; i++) {
            measureNode = measureNode.next;
            if (!node) {
                return dummy.next; // rreturn linked list
            }
        }

        let current = prev.next;
        let next = current.next; // next reference pointer for mid reversal
        for (let i = 1; i< k; i++) { // in place reversal
            current.next = next.next;
            next.next = prev.next
            prev.next = next
            next = current.next
        }

        prev = current;
    }

    return dummy.next
};